feat: ruff --statistics

`flake8` provides a nice `--statistics` flags which summarizes the findings at the end with a list like the following:

```
3     E126 continuation line over-indented for hanging indent
1     E225 missing whitespace around operator
3     E231 missing whitespace after ','
5     E701 multiple statements on one line (colon)
2     E711 comparison to None should be 'if cond is None:'
1     E714 test for object identity should be 'is not'
8     E999 SyntaxError: invalid syntax
40    F541 f-string is missing placeholders
5     F821 undefined name 'x'
```

It would be nice to have this as well.

`flake8 --statistics` doesn't play too well with custom formatters, and for example, produce invalid output with `--format=json`:

```console
$ flake8 --select B resources/test/fixtures/flake8_bugbear/B002.py --statistics
resources/test/fixtures/flake8_bugbear/B002.py:15:9: B002 Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.
resources/test/fixtures/flake8_bugbear/B002.py:20:12: B002 Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.
2     B002 Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.

$ flake8 --select B resources/test/fixtures/flake8_bugbear/B002.py --format json | jq
{
  "resources/test/fixtures/flake8_bugbear/B002.py": [
    {
      "code": "B002",
      "filename": "resources/test/fixtures/flake8_bugbear/B002.py",
      "line_number": 15,
      "column_number": 9,
      "text": "Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.",
      "physical_line": "    x = ++n\n"
    },
    {
      "code": "B002",
      "filename": "resources/test/fixtures/flake8_bugbear/B002.py",
      "line_number": 20,
      "column_number": 12,
      "text": "Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.",
      "physical_line": "    return ++n\n"
    }
  ]
}

$ flake8 --select B resources/test/fixtures/flake8_bugbear/B002.py --statistics --format json
{"resources/test/fixtures/flake8_bugbear/B002.py": [{"code": "B002", "filename": "resources/test/fixtures/flake8_bugbear/B002.py", "line_number": 15, "column_number": 9, "text": "Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.", "physical_line": "    x = ++n\n"}, {"code": "B002", "filename": "resources/test/fixtures/flake8_bugbear/B002.py", "line_number": 20, "column_number": 12, "text": "Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.", "physical_line": "    return ++n\n"}]2     B002 Python does not support the unary prefix increment. Writing ++n is equivalent to +(+(n)), which equals n. You meant n += 1.}%
```

`ruff --statistics` could improve this by sending the stats summary to stderr, so that the output can be safely piped.

--- a/ruff_cli/src/args.rs
+++ b/ruff_cli/src/args.rs
@@ -246,6 +246,9 @@ pub struct CheckArgs {
         conflicts_with = "watch",
     )]
     pub show_settings: bool,
+    /// Format and print the statistics
+    #[arg(long)]
+    pub statistics: bool,
 }
 
 #[derive(Debug, Clone, Copy, clap::ValueEnum)]
@@ -316,6 +319,7 @@ impl CheckArgs {
                 no_cache: self.no_cache,
                 show_files: self.show_files,
                 show_settings: self.show_settings,
+                statistics: self.statistics,
                 stdin_filename: self.stdin_filename,
                 watch: self.watch,
             },
@@ -371,6 +375,7 @@ pub struct Arguments {
     pub no_cache: bool,
     pub show_files: bool,
     pub show_settings: bool,
+    pub statistics: bool,
     pub stdin_filename: Option<PathBuf>,
     pub watch: bool,
 }

--- a/ruff_cli/src/args.rs
+++ b/ruff_cli/src/args.rs
@@ -246,7 +246,7 @@ pub struct CheckArgs {
         conflicts_with = "watch",
     )]
     pub show_settings: bool,
-    /// Format and print the statistics
+    /// Show counts for every rule with at least one violation.
     #[arg(long)]
     pub statistics: bool,
 }

--- a/ruff_cli/src/args.rs
+++ b/ruff_cli/src/args.rs
@@ -211,6 +211,15 @@ pub struct CheckArgs {
     update_check: bool,
     #[clap(long, overrides_with("update_check"), hide = true)]
     no_update_check: bool,
+    /// Show counts for every rule with at least one violation.
+    #[arg(
+        long,
+        // Unsupported default-command arguments.
+        conflicts_with = "diff",
+        conflicts_with = "show_source",
+        conflicts_with = "watch",
+    )]
+    pub statistics: bool,
     /// Enable automatic additions of `noqa` directives to failing lines.
     #[arg(
         long,
@@ -218,6 +227,7 @@ pub struct CheckArgs {
         conflicts_with = "show_files",
         conflicts_with = "show_settings",
         // Unsupported default-command arguments.
+        conflicts_with = "statistics",
         conflicts_with = "stdin_filename",
         conflicts_with = "watch",
     )]
@@ -230,6 +240,7 @@ pub struct CheckArgs {
         // conflicts_with = "show_files",
         conflicts_with = "show_settings",
         // Unsupported default-command arguments.
+        conflicts_with = "statistics",
         conflicts_with = "stdin_filename",
         conflicts_with = "watch",
     )]
@@ -242,13 +253,11 @@ pub struct CheckArgs {
         conflicts_with = "show_files",
         // conflicts_with = "show_settings",
         // Unsupported default-command arguments.
+        conflicts_with = "statistics",
         conflicts_with = "stdin_filename",
         conflicts_with = "watch",
     )]
     pub show_settings: bool,
-    /// Show counts for every rule with at least one violation.
-    #[arg(long)]
-    pub statistics: bool,
 }
 
 #[derive(Debug, Clone, Copy, clap::ValueEnum)]

--- a/ruff_cli/src/main.rs
+++ b/ruff_cli/src/main.rs
@@ -244,7 +244,11 @@ fn check(args: CheckArgs, log_level: LogLevel) -> Result<ExitCode> {
         // unless we're writing fixes via stdin (in which case, the transformed
         // source code goes to stdout).
         if !(is_stdin && matches!(autofix, fix::FixMode::Apply | fix::FixMode::Diff)) {
-            printer.write_once(&diagnostics)?;
+            if cli.statistics {
+                printer.write_statistics(&diagnostics)?;
+            } else {
+                printer.write_once(&diagnostics)?;
+            }
         }
 
         // Check for updates if we're in a non-silent log level.

--- a/ruff_cli/src/printer.rs
+++ b/ruff_cli/src/printer.rs
@@ -351,40 +351,60 @@ impl<'a> Printer<'a> {
             .collect::<Vec<_>>();
         violations.sort();
         violations.dedup();
-        let statistics = violations.iter().map(|rule| ExpandedStatistics {
-            code: rule.code(),
-            count: diagnostics
-                .messages
-                .iter()
-                .filter(|message| message.kind.rule() == *rule)
-                .count(),
-            message: diagnostics
-                .messages
-                .iter()
-                .find(|message| message.kind.rule() == *rule)
-                .map(|message| message.kind.body())
-                .unwrap(),
-        });
+
+        let statistics = violations
+            .iter()
+            .map(|rule| ExpandedStatistics {
+                code: rule.code(),
+                count: diagnostics
+                    .messages
+                    .iter()
+                    .filter(|message| message.kind.rule() == *rule)
+                    .count(),
+                message: diagnostics
+                    .messages
+                    .iter()
+                    .find(|message| message.kind.rule() == *rule)
+                    .map(|message| message.kind.body())
+                    .unwrap(),
+            })
+            .collect::<Vec<_>>();
 
         let mut stdout = BufWriter::new(io::stdout().lock());
         match self.format {
-            SerializationFormat::Json => {
-                writeln!(
-                    stdout,
-                    "{}",
-                    serde_json::to_string_pretty(&statistics.collect::<Vec<_>>())?
-                )?;
-            }
             SerializationFormat::Text => {
+                // Compute the maximum number of digits in the count and code, for all messages, to enable
+                // pretty-printing.
+                let count_width = num_digits(
+                    statistics
+                        .iter()
+                        .map(|statistic| statistic.count)
+                        .max()
+                        .unwrap(),
+                );
+                let code_width = statistics
+                    .iter()
+                    .map(|statistic| statistic.code.len())
+                    .max()
+                    .unwrap();
+
+                // By default, we mimic Flake8's `--statistics` format.
                 for msg in statistics {
-                    // same format as flake8 --statistics has
-                    writeln!(stdout, "{:<5} {} {}", msg.count, msg.code, msg.message)?;
+                    writeln!(
+                        stdout,
+                        "{:>count_width$}  {:<code_width$}  {}",
+                        msg.count, msg.code, msg.message
+                    )?;
                 }
                 return Ok(());
             }
+            SerializationFormat::Json => {
+                writeln!(stdout, "{}", serde_json::to_string_pretty(&statistics)?)?;
+            }
             _ => {
                 anyhow::bail!(
-                    "Unsuported format for --statistics! Allowed formats are text and json."
+                    "Unsupported serialization format for statistics: {:?}",
+                    self.format
                 )
             }
         }

--- a/ruff_cli/src/printer.rs
+++ b/ruff_cli/src/printer.rs
@@ -392,7 +392,7 @@ impl<'a> Printer<'a> {
                 for msg in statistics {
                     writeln!(
                         stdout,
-                        "{:>count_width$}  {:<code_width$}  {}",
+                        "{:>count_width$}\t{:<code_width$}\t{}",
                         msg.count, msg.code, msg.message
                     )?;
                 }

--- a/ruff_cli/src/printer.rs
+++ b/ruff_cli/src/printer.rs
@@ -43,6 +43,13 @@ struct ExpandedMessage<'a> {
     filename: &'a str,
 }
 
+#[derive(Serialize)]
+struct ExpandedStatistics<'a> {
+    count: usize,
+    code: &'a str,
+    message: String,
+}
+
 struct SerializeRuleAsCode<'a>(&'a Rule);
 
 impl Serialize for SerializeRuleAsCode<'_> {
@@ -328,6 +335,57 @@ impl<'a> Printer<'a> {
                     );
                     writeln!(stdout, "{label}")?;
                 }
+            }
+        }
+
+        stdout.flush()?;
+
+        Ok(())
+    }
+
+    pub fn write_statistics(&self, diagnostics: &Diagnostics) -> Result<()> {
+        let mut violations = diagnostics
+            .messages
+            .iter()
+            .map(|message| message.kind.rule())
+            .collect::<Vec<_>>();
+        violations.sort();
+        violations.dedup();
+        let statistics = violations.iter().map(|rule| ExpandedStatistics {
+            code: rule.code(),
+            count: diagnostics
+                .messages
+                .iter()
+                .filter(|message| message.kind.rule() == *rule)
+                .count(),
+            message: diagnostics
+                .messages
+                .iter()
+                .find(|message| message.kind.rule() == *rule)
+                .map(|message| message.kind.body())
+                .unwrap(),
+        });
+
+        let mut stdout = BufWriter::new(io::stdout().lock());
+        match self.format {
+            SerializationFormat::Json => {
+                writeln!(
+                    stdout,
+                    "{}",
+                    serde_json::to_string_pretty(&statistics.collect::<Vec<_>>())?
+                )?;
+            }
+            SerializationFormat::Text => {
+                for msg in statistics {
+                    // same format as flake8 --statistics has
+                    writeln!(stdout, "{:<5} {} {}", msg.count, msg.code, msg.message)?;
+                }
+                return Ok(());
+            }
+            _ => {
+                anyhow::bail!(
+                    "Unsuported format for --statistics! Allowed formats are text and json."
+                )
             }
         }
 

--- a/ruff_cli/tests/integration_test.rs
+++ b/ruff_cli/tests/integration_test.rs
@@ -196,7 +196,7 @@ fn show_statistics() -> Result<()> {
             .lines()
             .last()
             .unwrap(),
-        "1     F401 `sys` imported but unused"
+        "1\tF401\t`sys` imported but unused"
     );
     Ok(())
 }

--- a/ruff_cli/tests/integration_test.rs
+++ b/ruff_cli/tests/integration_test.rs
@@ -182,3 +182,21 @@ fn explain_status_codes() -> Result<()> {
     cmd.args(["--explain", "RUF404"]).assert().failure();
     Ok(())
 }
+
+#[test]
+fn show_statistics() -> Result<()> {
+    let mut cmd = Command::cargo_bin(BIN_NAME)?;
+    let output = cmd
+        .args(["-", "--format", "text", "--select", "F401", "--statistics"])
+        .write_stdin("import sys\nimport os\n\nprint(os.getuid())\n")
+        .assert()
+        .failure();
+    assert_eq!(
+        str::from_utf8(&output.get_output().stdout)?
+            .lines()
+            .last()
+            .unwrap(),
+        "1     F401 `sys` imported but unused"
+    );
+    Ok(())
+}

